wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java

```diff
@@ -0,0 +1,117 @@
+  public static final ProtoTypeName FIXED64 = new ProtoTypeName("fixed64");
+  public static final ProtoTypeName INT32 = new ProtoTypeName("int32");
+  public static final ProtoTypeName INT64 = new ProtoTypeName("int64");
+  public static final ProtoTypeName SFIXED32 = new ProtoTypeName("sfixed32");
+  public static final ProtoTypeName SFIXED64 = new ProtoTypeName("sfixed64");
+  public static final ProtoTypeName SINT32 = new ProtoTypeName("sint32");
+  public static final ProtoTypeName SINT64 = new ProtoTypeName("sint64");
+  public static final ProtoTypeName STRING = new ProtoTypeName("string");
+  public static final ProtoTypeName UINT32 = new ProtoTypeName("uint32");
+  public static final ProtoTypeName UINT64 = new ProtoTypeName("uint64");
+
+  private static final Map<String, ProtoTypeName> SCALAR_TYPES;
+  static {
+    Map<String, ProtoTypeName> scalarTypes = new LinkedHashMap<String, ProtoTypeName>();
+    try {
+      for (Field field : ProtoTypeName.class.getDeclaredFields()) {
+        if (field.getType() == ProtoTypeName.class) {
+          ProtoTypeName scalar = (ProtoTypeName) field.get(null);
+          scalarTypes.put(scalar.names.get(0), scalar);
+        }
+      }
+      SCALAR_TYPES = Collections.unmodifiableMap(scalarTypes);
+    } catch (IllegalAccessException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private final String protoPackage;
+  private final List<String> names;
+  private final boolean isScalar;
+
+  private ProtoTypeName(String scalarName) {
+    this(null, Collections.singletonList(scalarName), true);
+  }
+
+  private ProtoTypeName(String protoPackage, List<String> names, boolean isScalar) {
+    this.protoPackage = protoPackage;
+    this.names = names;
+    this.isScalar = isScalar;
+  }
+
+  public String simpleName() {
+    return names.get(names.size() - 1);
+  }
+
+  public static ProtoTypeName get(String protoPackage, String name) {
+    checkNotNull(name, "name");
+    return new ProtoTypeName(protoPackage, Collections.singletonList(name), false);
+  }
+
+  public static ProtoTypeName getScalar(String name) {
+    return SCALAR_TYPES.get(name);
+  }
+
+  public ProtoTypeName nestedType(String name) {
+    checkState(!isScalar);
+    checkNotNull(name, "name");
+    return new ProtoTypeName(protoPackage, Util.concatenate(names, name), false);
+  }
+
+  @Override public boolean equals(Object o) {
+    return o instanceof ProtoTypeName
+        && Util.equal(((ProtoTypeName) o).protoPackage, protoPackage)
+        && ((ProtoTypeName) o).names.equals(names)
+        && ((ProtoTypeName) o).isScalar == isScalar;
+  }
+
+  @Override public int hashCode() {
+    return super.hashCode();
+  }
+
+  @Override public String toString() {
+    StringBuilder result = new StringBuilder();
+    if (protoPackage != null) {
+      result.append(protoPackage);
+      result.append(".");
+    }
+    for (int i = 0; i < names.size(); i++) {
```

Comment:
Who is this person who wrote the non-cached size for loop? I don't even know him!